### ISSUE TITLE
How to analyse on new dataset?

### ISSUE THREAD

**Issue by yingmuzhi** [LeonardoLupori/wholeBrain_PNN_analysis#2]
Hello, we have run your script, and it's great. We expect to apply your data analysis code to our own new dataset for the analysis of WFA. However, we don't know how to use it. Taking AL1A as an example, in your data, the `cellID, parentImg, x, y, xCCF, yCCF, zCCF, regionID, fluoMean, fluoMedian, areaPx` in the dot file, the `regionID, areaPx, areaMm2, diffFluo, avgPxIntensity` in the diffFluo file, and the `cellID, parentImg, x, y, xCCF, yCCF, zCCF, regionID, wfa, fluoMeanWfa, fluoMedianWfa, areaPxWfa, pv, fluoMeanPv, fluoMedianPv, areaPxPv` in the combinedDots file are statistically analyzed. How were these amounts of data obtained?
Waiting for your patient reply, thanks :)

**Comment by yingmuzhi** [comment]
We have carefully read your paper. In the "Method" section, you mentioned that your csv data was obtained through a Matlab program. We would like to know how to obtain the three files: `AL1A_diffFluo_wfa_20220412-154156.csv`, `AL1A_dots_wfa_20220623-084430.csv` and `AL1A_combinedDots.csv`. We would like to use your excellent analysis program to analyze new data and hope you can provide some help. Thank you very much.
Waiting for your patient reply, thanks :)

**Comment by LeonardoLupori** [comment]
Thanks a lot for the question

To get a sense of how we went from raw to preprocessed data you could take a look at this [repo](https://github.com/LeonardoLupori/brainAlignment). It's not as polished as the one to reproduce all the figures since the majority of the analysis is relatively specific for the way in which we organize data.
In particular there are 2 functions that could be very useful to read through. One generates preprocessed data for diffuse fluorescence and the other for cell intensity and cell count for a series of slices from one mouse:

1. for diffuse fluorescence[ here](https://github.com/LeonardoLupori/brainAlignment/blob/main/A2_batchQuantifyDiffuse.m)
2. for "dots" (meaning cells or PNNs - we do not use the word "cells" since PNNs are not cells) see[ here](https://github.com/LeonardoLupori/brainAlignment/blob/main/A1_batchQuantifyDots.m)

It's also probably worth taking a look at the Slice class[ here](https://github.com/LeonardoLupori/brainAlignment/blob/main/Slice.m). This is meant to represent a single brain slice and contains information about the alignment. The methods of this class are used to quantify its fluorescence and cell-level metrics.

This set of functions is meant to work with data structured in a relatively stringent way. You can actually download the entire dataset of raw images used in the paper and all the relative annotations [here](https://zenodo.org/records/7419283) to take a look. Briefly, each mouse has its folder containing raw images, subsampled images, cell counts, alignment info.

To sum up, if you want to reproduce the visualizations with your data the easiest solution would be to copy the data structure in our raw data. Then you can use the 2 batch functions to calculate preprocessed data and from that use the main github repo to visualize the data.

One consideration: when quantifying "dots" we use a simple random forest to segment background and foreground for every cell. This is done to get an estimate of the cell fluorescence that is as accurate as possible by averaging only pixels that belong to the cell/PNN. You see the model provided at[ this line](https://github.com/LeonardoLupori/brainAlignment/blob/8eec67dcb1fed7bbea021edcb92eb501a9e52351/Slice.m#L438) for example. However it is almost certain that you will need to train a custom model specific for the staining that you want to classify. For that take a look at[ our implementation](https://github.com/LeonardoLupori/brainAlignment/blob/main/cellClassifier.m) if you want. You can also use any other segmentation algorithm around, like cellpose for example.

Let me know if you have any other questions and I'd be happy to help.

best, Leo

**Comment by yingmuzhi** [comment]
Thank you very much for your reply!! Leo!!

It will take me some time to read Matlab program. 
It is very nice of you to help me, Hoping you will have a nice day. 

Best regards, Max.

**Comment by yingmuzhi** [comment]
Hello, Leo. 

I have carefully gone through your repository. However, when I ran the `A2_batchQuantifyDiffuse.m` file, I encountered the following problem as shown in Figure 1. It prompted that the `annotationVolume.mat` file was not found.

![Image](https://github.com/user-attachments/assets/d0340ea8-ff35-475d-b64e-ebfe09f188b7)

We used the data of AL1A in your rawData, and it was created strictly in accordance with the requirements of the repository, as shown in Figure 2 below.

![Image](https://github.com/user-attachments/assets/7175be86-83c4-4306-80db-b53d5ac78d78)

It seems that `annotationVolume.mat` file was lost. Waiting for your patient reply, thanks :)

best regards, max.

**Comment by LeonardoLupori** [comment]
The missing file is the annotation volume of the Allen institute. You can download it using ther API:

docs  [here](https://allensdk.readthedocs.io/en/latest/reference_space.html)
example [here](https://allensdk.readthedocs.io/en/latest/_static/examples/nb/reference_space.html#Downloading-an-annotation-volume)

or, in a Python environment where you have allensdk installed:
```python
from pathlib import Path
import os

from allensdk.core.reference_space_cache import ReferenceSpaceCache

output_dir = '.'

reference_space_key = os.path.join('annotation', 'ccf_2017')
resolution = 25
rspc = ReferenceSpaceCache(resolution, reference_space_key, manifest=Path(output_dir) / 'manifest.json')
# ID 1 is the adult mouse structure graph
tree = rspc.get_structure_tree(structure_graph_id=1) 

annotation, meta = rspc.get_annotation_volume()
# The file should appear in the reference space key directory
os.listdir(Path(output_dir) / reference_space_key)
```

The volume you get here is a .nrrd file that in Matlab requires a paid toolbox (medical imaging) to read, or this custom function some saint wrote (https://ch.mathworks.com/matlabcentral/fileexchange/50830-nrrd-format-file-reader). 

I saved it as a .mat file for convenience. You can find it here [anotationVolume.zip](https://github.com/user-attachments/files/20203281/anotationVolume.zip)
For convenience, I saved it so that in the 3rd dimension you can go through different coronal planes. This means that before feeding the variable to the pipeline, you need to do some transformation though. here's how:

```matlab
load("path/to/annotationVolume.mat")
% switch the order of dimensions
annotationVolume = permute(annotationVolume, [2,3,1]);
% Flip 
annotationVolume = annotationVolume(end:-1:1, end:-1:1, end:-1:1);


%to test if the volume works fine I would load a single slice from the AL1A mouse and get the oblique slice of the volume with:

sl = Slice(sliceNumber, "path/to/xmInfoFile.xml");
obliqueSlice = sl.getObliqueSlice(annotationVolume );

% compare the slice thumbnail with the oblique slice of the volume. They should match very closely (except local deformations that are yet to be applied from the displacement field)

subplot 121
% rank the annotation values, otherwise the colormap is skewed for extremely large region IDs
[~, ~, ranks] = unique(obliqueSlice(:));            
ranked = reshape(ranks, size(obliqueSlice));
imagesc(ranked)
subplot 122
imagesc(sl.thumbnail(:,:,1))
```


this is the output I get for example for slice 88 of mouse AL1A

![Image](https://github.com/user-attachments/assets/df57a052-711c-4784-9d3e-3588d21a4cbe)

let me know if it helps.

**Comment by yingmuzhi** [comment]
Hello, Leo. 

Thank you very much for explaining it to me in such detail and giving such a good example in both Python and Matlab ways! We successfully ran the file `A2_batchQuantifyDiffuse.m`. And output `AL1A_diffFluo_pv_20250515-092532.csv` and `AL1A_sliceFluo_pv_20250515-092532.csv` 👍 

When we were running the file `A1_batchQuantifyDots.m`, it failed to find the file `model_pv_20220412-1829.mat`. Combined with your previous reply, I think this might be that I didn't run `cellClassifier.m` and generate the corresponding model. It would be of great help to me if you could demonstrate how to run the random forest algorithm using `cellClassifier.m` on the AL1A dataset and generate `model_pv_20220412-1829.mat`.

Besides, this question of mine might be rather stupid. As far as I know, the random forest algorithm is often used for classification and regression, but here it seems that you have used it for segmentation? 

I'm really looking forward to your patient reply : )

best regards, max.

**Comment by LeonardoLupori** [comment]
> When we were running the file `A1_batchQuantifyDots.m`, it failed to find the file `model_pv_20220412-1829.mat`. Combined with your previous reply, I think this might be that I didn't run `cellClassifier.m` and generate the corresponding model. It would be of great help to me if you could demonstrate how to run the random forest algorithm using `cellClassifier.m` on the AL1A dataset and generate `model_pv_20220412-1829.mat`.
> 

For this testing stage we can share the models we used for segmentation of PV and WFA channels. I do not have the files at the moment but @valtot could share them with you here. He could also provide a bit of guidance on how to label your training set for your images and do the segmentations yourself :)


> Besides, this question of mine might be rather stupid. As far as I know, the random forest algorithm is often used for classification and regression, but here it seems that you have used it for segmentation?

No such thing as stupid questions :)
This is correct. It is commonly used for classification. Our implementation does pixel-wise classification of foreground/background. If you do it for every pixel in a small image crop around the neuron, then you end up with a image of FG/BG labels that correspond to a segmentation mask of the cell pixels.

**Comment by yingmuzhi** [comment]
Hello, Leo.

Very good code, with strong reproducibility, and open-source is also well done, it should be the industry benchmark in counting PNNs in brains.

But I am still encountering some issues. After successfully running P1-P6, deep learning model, rf01, rf02, a1, a2, I obtained the `sliceFluo.csv`, `dots.csv`, and `diffFluo.csv` files for pv and wfa, but I did not obtain the `combinedDots.csv` file. I would like to know what its logic is, does it refer to the statistics of the same xy wfa and pv?

I'm really looking forward to your patient reply : )

best regards, max.

**Comment by valtot** [comment]
Hi!
Thank you for pointing this out! The `combinedDots.csv` file is used for analyzing colocalization between WFA+ and PV+ cells. All the cells detected with either staining in an mouse dataset are here collected with the indication of their fluorescence in the WFA and/or the PV channel. I have now added in the [brainAlignment ](https://github.com/LeonardoLupori/brainAlignment/commit/148b95103943b074fb9ce2ce4f27bed392c8c2b5) repo the code that is used to produce such files. Please note that in the current version of the code only WFA/PV double staining is supported.

Hope this helps!
